NullPointerException in KafkaClient#getGroupOffsets

Hi,
As I'm facing to Invalid negative offset issue, I update kafka-lag-exporter from 0.6.0 to 0.6.2. 

On startup, an error occurred and no Kafka metric is exported.
`NullPointerException` on `com.lightbend.kafkalagexporter.KafkaClient.$anonfun$getGroupOffsets$7(KafkaClient.scala:256)`

Full stacktrace

```
2020-06-24T14:28:56.961396971Z 2020-06-24 14:28:56,961 INFO  o.a.kafka.common.utils.AppInfoParser akka://kafka-lag-exporter/user/consumer-group-collector-xxx - Kafka version: 2.5.0 
2020-06-24T14:28:56.961415253Z 2020-06-24 14:28:56,961 INFO  o.a.kafka.common.utils.AppInfoParser akka://kafka-lag-exporter/user/consumer-group-collector-xxx - Kafka commitId: 66563e712b0b9f84 
2020-06-24T14:28:56.961474391Z 2020-06-24 14:28:56,961 INFO  o.a.kafka.common.utils.AppInfoParser akka://kafka-lag-exporter/user/consumer-group-collector-xxx - Kafka startTimeMs: 1593008936960 
2020-06-24T14:28:59.399461538Z 2020-06-24 14:28:59,398 INFO  org.apache.kafka.clients.Metadata  - [Consumer clientId=kafla-consumer-lag-exporter, groupId=kafka-consumer-lag-exporter] Cluster ID: xxx
2020-06-24T14:28:59.802002245Z 2020-06-24 14:28:59,801 ERROR c.l.k.ConsumerGroupCollector$ akka://kafka-lag-exporter/user/consumer-group-collector-xxx - Supervisor RestartSupervisor saw failure: A failure occurred while retrieving offsets.  Shutting down. java.lang.Exception: A failure occurred while retrieving offsets.  Shutting down.
2020-06-24T14:28:59.802061065Z 	at com.lightbend.kafkalagexporter.ConsumerGroupCollector$CollectorBehavior.$anonfun$collector$1(ConsumerGroupCollector.scala:201)
2020-06-24T14:28:59.802068282Z 	at akka.actor.typed.internal.BehaviorImpl$ReceiveBehavior.receive(BehaviorImpl.scala:136)
2020-06-24T14:28:59.80207228Z 	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
2020-06-24T14:28:59.802076351Z 	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
2020-06-24T14:28:59.802080262Z 	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:55)
2020-06-24T14:28:59.802084151Z 	at akka.actor.typed.internal.RestartSupervisor.aroundReceive(Supervision.scala:262)
2020-06-24T14:28:59.802087735Z 	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:83)
2020-06-24T14:28:59.802091541Z 	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
2020-06-24T14:28:59.802095091Z 	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
2020-06-24T14:28:59.80209865Z 	at akka.actor.typed.internal.adapter.ActorAdapter.handleMessage(ActorAdapter.scala:125)
2020-06-24T14:28:59.802102371Z 	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:105)
2020-06-24T14:28:59.802106148Z 	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:573)
2020-06-24T14:28:59.802109831Z 	at akka.actor.ActorCell.invoke(ActorCell.scala:543)
2020-06-24T14:28:59.80211335Z 	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:269)
2020-06-24T14:28:59.802116893Z 	at akka.dispatch.Mailbox.run(Mailbox.scala:230)
2020-06-24T14:28:59.802120535Z 	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)
2020-06-24T14:28:59.802124084Z 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
2020-06-24T14:28:59.80212771Z 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
2020-06-24T14:28:59.802131388Z 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
2020-06-24T14:28:59.802135011Z 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
2020-06-24T14:28:59.802138781Z Caused by: java.lang.NullPointerException: null
2020-06-24T14:28:59.802162404Z 	at com.lightbend.kafkalagexporter.KafkaClient.$anonfun$getGroupOffsets$7(KafkaClient.scala:256)
2020-06-24T14:28:59.802166944Z 	at com.lightbend.kafkalagexporter.KafkaClient.$anonfun$getGroupOffsets$7$adapted(KafkaClient.scala:256)
2020-06-24T14:28:59.802170204Z 	at scala.Option.map(Option.scala:230)
2020-06-24T14:28:59.802173248Z 	at com.lightbend.kafkalagexporter.KafkaClient.$anonfun$getGroupOffsets$6(KafkaClient.scala:256)
2020-06-24T14:28:59.802176417Z 	at scala.collection.immutable.List.flatMap(List.scala:338)
2020-06-24T14:28:59.802180235Z 	at com.lightbend.kafkalagexporter.KafkaClient.getGroupOffsets(KafkaClient.scala:255)
2020-06-24T14:28:59.802183351Z 	at com.lightbend.kafkalagexporter.KafkaClient.$anonfun$getGroupOffsets$3(KafkaClient.scala:225)
2020-06-24T14:28:59.802186495Z 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
2020-06-24T14:28:59.802189534Z 	at scala.util.Success.map(Try.scala:213)
2020-06-24T14:28:59.802192551Z 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
2020-06-24T14:28:59.802195649Z 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
2020-06-24T14:28:59.802198715Z 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
2020-06-24T14:28:59.802201804Z 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
2020-06-24T14:28:59.80220487Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2020-06-24T14:28:59.802207991Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2020-06-24T14:28:59.802211153Z 	at java.lang.Thread.run(Thread.java:748)
```

Thanks for your help !

--- a/src/main/scala/com/lightbend/kafkalagexporter/KafkaClient.scala
+++ b/src/main/scala/com/lightbend/kafkalagexporter/KafkaClient.scala
@@ -253,8 +253,15 @@ class KafkaClient private[kafkalagexporter](cluster: KafkaCluster,
                                                 offsetMap: Map[KafkaTopicPartition, OffsetAndMetadata]): GroupOffsets =
     (for {
       gtp <- gtps
-      offset <- offsetMap.get(gtp.tp.asKafka).map(_.offset())
-    } yield gtp -> Some(LookupTable.Point(offset, now))).toMap
+      offsetResult <- offsetMap.get(gtp.tp.asKafka)
+    } yield {
+      // Offset can be null if it's invalid.  See javadocs:
+      // https://kafka.apache.org/25/javadoc/org/apache/kafka/clients/admin/ListConsumerGroupOffsetsResult.html#partitionsToOffsetAndMetadata--
+      if (offsetResult == null)
+        gtp -> None
+      else
+        gtp -> Some(LookupTable.Point(offsetResult.offset(), now))
+    }).toMap
 
   def close(): Unit = {
     adminClient.close()

--- a/src/test/scala/com/lightbend/kafkalagexporter/KafkaClientSpec.scala
+++ b/src/test/scala/com/lightbend/kafkalagexporter/KafkaClientSpec.scala
@@ -10,7 +10,7 @@ import com.lightbend.kafkalagexporter.Domain.GroupOffsets
 import com.lightbend.kafkalagexporter.KafkaClient.{AdminKafkaClientContract, ConsumerKafkaClientContract, KafkaTopicPartitionOps}
 import org.apache.kafka.clients.admin._
 import org.apache.kafka.clients.consumer.OffsetAndMetadata
-import org.apache.kafka.common.{ConsumerGroupState, TopicPartition => TP}
+import org.apache.kafka.common.{ConsumerGroupState, TopicPartition => KafkaTopicPartition}
 import org.mockito.{ArgumentMatchersSugar, MockitoSugar}
 import org.scalatest.concurrent.ScalaFutures
 import org.scalatest.{FreeSpec, Matchers}
@@ -86,6 +86,22 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
         groupOffsets(gtp0_0) shouldEqual Some(LookupTable.Point(0, 0))
         groupOffsets(gtp0_1) shouldEqual Some(LookupTable.Point(1, 0))
       }
+
+      "returns no offsets when a null offset is returned" in {
+        val groups = List(groupId)
+        val gtps = List(gtp0)
+
+        val (adminKafkaClient, _, client) = getClient()
+
+        val listGroupOffsetResults = Future.successful(Map[KafkaTopicPartition, OffsetAndMetadata](
+          topicPartition0.asKafka -> null
+        ).asJava)
+        when(adminKafkaClient.listConsumerGroupOffsets(groupId)).thenReturn(listGroupOffsetResults)
+
+        val groupOffsets = client.getGroupOffsets(0, groups, gtps).futureValue
+
+        groupOffsets(gtp0) shouldEqual None
+      }
     }
 
     "getOffsetOrZero returns offsets of None (Option[Point]) for missing partitions" in {
@@ -107,7 +123,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       val tmpHost = "brokers"
       "will default to fetching all topics" in {
         val (_,_,client) = getClient()
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)
@@ -118,7 +134,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       "will only fetch whitelisted topic when whitelist contains a single topic" in {
         val tmpCluster = cluster.copy(topicWhitelist = List(topic2))
         val (_,_,client) = getClient(tmpCluster)
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)
@@ -129,7 +145,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       "will only fetch non blacklisted topics" in {
         val tmpCluster: KafkaCluster = cluster.copy(topicWhitelist = List(topic, topic2), topicBlacklist = List(topic))
         val (_,_,client) = getClient(tmpCluster)
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)
@@ -140,7 +156,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       "will only fetch whitelisted topics when whitelist is a regex against multiple topics" in {
         val tmpCluster = cluster.copy(topicWhitelist = List("test.+"))
         val (_,_,client) = getClient(tmpCluster)
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)
@@ -151,7 +167,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       "will exclude topics that match with blacklist regex" in {
         val tmpCluster = cluster.copy(topicWhitelist = List("test.+"), topicBlacklist = List(".+topic-2"))
         val (_,_,client) = getClient(tmpCluster)
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)
@@ -162,7 +178,7 @@ class KafkaClientSpec extends FreeSpec with Matchers with TestData with MockitoS
       "will not return any topics if the whitelist is empty" in {
         val tmpCluster = cluster.copy(topicWhitelist = List.empty)
         val (_,_,client) = getClient(tmpCluster)
-        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new TP(tp.topic, tp.partition)).asJava
+        val tps = Set(topicPartition0, topicPartition1, topicPartition2, topic2Partition0).map(tp => new KafkaTopicPartition(tp.topic, tp.partition)).asJava
         val members = List(new MemberDescription(consumerId, clientId, tmpHost, new MemberAssignment(tps))).asJava
 
         val description = new ConsumerGroupDescription(groupId, true, members, "", ConsumerGroupState.STABLE, node)

